src/internal/streamstore/streamstore.go

```diff
@@ -99,3 +104,143 @@ func (di *streamItem) ToReader() io.ReadCloser {
 func (di *streamItem) Deleted() bool {
 	return false
 }
+
+// ---------------------------------------------------------------------------
+// common reader/writer/deleter
+// ---------------------------------------------------------------------------
+
+type backuper interface {
+	BackupCollections(
+		ctx context.Context,
+		bases []kopia.IncrementalBase,
+		cs []data.BackupCollection,
+		globalExcludeSet map[string]map[string]struct{},
+		tags map[string]string,
+		buildTreeWithBase bool,
+		errs *fault.Bus,
+	) (*kopia.BackupStats, *details.Builder, map[string]kopia.PrevRefs, error)
+}
+
+// write persists bytes to the store
+func write(
+	ctx context.Context,
+	tenantID string,
+	service path.ServiceType,
+	collectionPurpose string,
+	itemName string,
+	bup backuper,
+	mr Marshaller,
+	errs *fault.Bus,
+) (string, error) {
+	// construct the path of the container
+	p, err := path.Builder{}.
+		ToStreamStorePath(tenantID, collectionPurpose, service, false)
+	if err != nil {
+		return "", clues.Stack(err).WithClues(ctx)
+	}
+
+	// TODO: We could use an io.Pipe here to avoid a double copy but that
+	// makes error handling a bit complicated
+	bs, err := mr.Marshal()
+	if err != nil {
+		return "", clues.Wrap(err, "marshalling body").WithClues(ctx)
+	}
+
+	dc := &streamCollection{
+		folderPath: p,
+		item: &streamItem{
+			name: itemName,
+			data: bs,
+		},
+	}
+
+	backupStats, _, _, err := bup.BackupCollections(
+		ctx,
+		nil,
+		[]data.BackupCollection{dc},
+		nil,
+		nil,
+		false,
+		errs)
+	if err != nil {
+		return "", errors.Wrap(err, "storing marshalled bytes in repository")
+	}
+
+	return backupStats.SnapshotID, nil
+}
+
+type restorer interface {
+	RestoreMultipleItems(
+		ctx context.Context,
+		snapshotID string,
+		paths []path.Path,
+		bc kopia.ByteCounter,
+		errs *fault.Bus,
+	) ([]data.RestoreCollection, error)
+}
+
+// read retrieves an object from the store
+func read(
+	ctx context.Context,
+	tenantID string,
+	service path.ServiceType,
+	collectionPurpose string,
+	itemName string,
+	id string,
+	rer restorer,
+	umr Unmarshaller,
+	errs *fault.Bus,
+) error {
+	// construct the path of the container
+	p, err := path.Builder{}.
+		Append(itemName).
+		ToStreamStorePath(tenantID, collectionPurpose, service, true)
+	if err != nil {
+		return clues.Stack(err).WithClues(ctx)
+	}
+
+	cs, err := rer.RestoreMultipleItems(
+		ctx,
+		id,
+		[]path.Path{p},
+		&stats.ByteCounter{},
+		errs)
+	if err != nil {
+		return errors.Wrap(err, "retrieving data")
+	}
+
+	// Expect only 1 data collection
+	if len(cs) != 1 {
+		return clues.New("greater than 1 collection found").
```

Comment:
It could have found 0 collections ;)
```suggestion
		return clues.New("unexpected number of collections found").
```